velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,585 @@
+
+    for (int i = 0; i < size; ++i) {
+      const auto count = flatValues->valueAt(offset + i);
+      accumulator->addValueWithCount(*mapKeys, offset + i, count, allocator);
+    }
+  }
+
+  DecodedVector decodedKeys_;
+  DecodedVector decodedIntermediate_;
+  TypePtr inputType_;
+};
+} // namespace
+
+void registerModeAggregate(
+    const std::string& prefix,
+    bool withCompanionFunctions,
+    bool overwrite) {
+  std::vector<std::shared_ptr<AggregateFunctionSignature>> signatures = {
+      AggregateFunctionSignatureBuilder()
+          .typeVariable("T")
+          .returnType("T")
+          .intermediateType("map(T,bigint)")
+          .argumentType("T")
+          .build(),
+  };
+
+  auto name = prefix + "mode";
+  registerAggregateFunction(
+      name,
+      std::move(signatures),
+      [name](
+          core::AggregationNode::Step step,
+          const std::vector<TypePtr>& argTypes,
+          const TypePtr& resultType,
+          const core::QueryConfig& /*config*/) -> std::unique_ptr<Aggregate> {
+        VELOX_CHECK_EQ(
+            argTypes.size(), 1, "{}: unexpected number of arguments", name);
+
+        auto inputType =
+            isRawInput(step) ? argTypes[0] : argTypes[0]->childAt(0);
+        switch (inputType->kind()) {
+          case TypeKind::BOOLEAN:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<bool>>>(resultType);
+          case TypeKind::TINYINT:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<int8_t>>>(resultType);
+          case TypeKind::SMALLINT:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<int16_t>>>(resultType);
+          case TypeKind::INTEGER:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<int32_t>>>(resultType);
+          case TypeKind::BIGINT:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<int64_t>>>(resultType);
+          case TypeKind::HUGEINT:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<int128_t>>>(resultType);
+          case TypeKind::REAL:
+            return std::make_unique<SimpleAggregateAdapter<ModeAggregate<
+                float,
+                util::floating_point::NaNAwareHash<float>,
+                util::floating_point::NaNAwareEquals<float>>>>(resultType);
+          case TypeKind::DOUBLE:
+            return std::make_unique<SimpleAggregateAdapter<ModeAggregate<
+                double,
+                util::floating_point::NaNAwareHash<double>,
+                util::floating_point::NaNAwareEquals<double>>>>(resultType);
+          case TypeKind::TIMESTAMP:
+            return std::make_unique<
+                SimpleAggregateAdapter<ModeAggregate<Timestamp>>>(resultType);
+          case TypeKind::UNKNOWN:
+            return std::make_unique<SimpleAggregateAdapter<ModeAggregate<
```

Comment:
`HistogramAggregate` supports UNKNOWN, you could take it as example

Reply:
`HistogramAggregate` use `int8_t` to register for `UNKNOWN`. I'm little supprisied for this. I can' use it for Mode as if the types doesn't match, ti would trigger the NPE in SimpleAggregateAdapter:
https://github.com/facebookincubator/velox/blob/8ab9d11c95fbfd93d6381f95238adceef5109052/velox/exec/SimpleAggregateAdapter.h#L291

Reply:
Yeah, I known your implementation, can you add the explain to `UnknownValueHash`?